### PR TITLE
app, ecs_client: Terminal error handling for registerContainerInstance api

### PR DESCRIPTION
**Pull request by sharanyad**
### Summary
All the Server and Client exception errors from Register Container Instance call are transient errors.
Making InvalidParameterException error to be non-transient i.e, Agent will not restart on InvalidParameter error.

### Implementation details
Check if the AWS error returned from register container instance API is of the type InvalidParameterException and return it as non-transient error. 

### Testing
- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: yes

### Description for the changelog
Making invalid parameter exception error from register container instance API call non-transient

### Licensing
This contribution is under the terms of the Apache 2.0 License: yes

### CODE REVIEW

File: agent/app/agent.go [hunk 1]
```diff
@@ -402,6 +404,10 @@ func (agent *ecsAgent) registerContainerInstance(
 			seelog.Critical("Instance registration attempt with an invalid attribute")
 			return err
 		}
+		awsErr, ok := err.(awserr.Error)
+		if ok && awsErr.Code() == ecs_service.ErrCodeInvalidParameterException {
+			return err
```
Comment (aaithal) [inline agent/app/agent.go]: Can you also add a log line here? Similar to the `"invalid attribute"` one above?

File: agent/app/agent.go [hunk 2]
```diff
@@ -402,6 +404,10 @@ func (agent *ecsAgent) registerContainerInstance(
 			seelog.Critical("Instance registration attempt with an invalid attribute")
 			return err
 		}
+		awsErr, ok := err.(awserr.Error)
```
Comment (aaithal) [inline agent/app/agent.go]: This should be before `utils.AttributeError` comparison. Also, a minor nit here: `if awsErr, ok := ...` pattern leads to lesser locs.

File: agent/app/agent.go [hunk 3]
```diff
@@ -47,6 +47,8 @@ import (
 	aws_credentials "github.com/aws/aws-sdk-go/aws/credentials"
 	"github.com/aws/aws-sdk-go/aws/defaults"
 	"github.com/cihub/seelog"
+	"github.com/aws/aws-sdk-go/aws/awserr"
```
Comment (vsiddharth) [inline agent/app/agent.go]: nit: move these above the 3p packages

File: agent/app/agent.go (hunk 2 above)
Comment (sharanyad) [reply agent/app/agent.go]: done.

File: agent/app/agent.go (hunk 1 above)
Comment (sharanyad) [reply agent/app/agent.go]: done

File: agent/app/agent.go (hunk 3 above)
Comment (sharanyad) [reply agent/app/agent.go]: done

File: agent/app/agent.go [hunk 4]
```diff
@@ -46,6 +46,8 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
 	aws_credentials "github.com/aws/aws-sdk-go/aws/credentials"
 	"github.com/aws/aws-sdk-go/aws/defaults"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	ecs_service "github.com/aws/aws-sdk-go/service/ecs"
```
Comment (samuelkarp) [inline agent/app/agent.go]: Please use the private generated client that we have at `github.com/aws/amazon-ecs-agent/agent/ecs_client/model/ecs` instead.  I'd rather not vendor two copies of the client.  We *may* need to update the [generator](https://github.com/aws/amazon-ecs-agent/blob/v1.14.5/agent/gogenerate/awssdk.go) to generate these constants, but I would prefer that to vendoring an additional copy.

**Comment by sharanyad** [comment]
Generated error types in `ecs_client`, instead of using aws-sdk-go client vendor package. Please have a look again. Thanks!

**Pull request merged**